Shell add getopt function

jakub-uC opened the pull request:
@carlescufi: This implementation is based on FreeBSD. I've changed it a bit to support multi-instance usage.
However, because of that, it is not 1:1 usage as regular getopt. Can we live with that?

Commit:
shell: make detecting help option optional

Shell will not "steal" by default "-h" and "--help" each time
help functions are enabled.
This change is necessary to implement and use the getopt library.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
lib: add freebsd getopt library

This library is going to be used by the shell module. Some shell users
are not satisfied with subcommands alone and need to use the options
for commands as well.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: add getopt library support

This functionality is is enabled by setting CONFIG_SHELL_GETOPT.
It is not active by default.

User can call following functions inside command handlers:
 - shell_getopt - getopt function based on freebsd implementation
 - shell_getopt_status_get - returns getopt status

Beware when getopt functionality is enabled shell will not parse
command handler to look for "-h" or "--help" options and print
help message automatically.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
shell: sample: add getopt

Modify an example by adding command that is using shell_getopt funtion.
Extent sample.yaml with getopt configuration.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nordic-krch commented on include/shell/shell.h:
```diff
@@ -958,6 +963,38 @@ void shell_help(const struct shell *shell);
+ *
+ * @returns		If an option was successfully found, function returns
+ *			the option character.
+ * @returns		If options have been detected that is not in ostr
```
you can use `@p ostr`

nordic-krch commented on include/shell/shell.h:
```diff
@@ -958,6 +963,38 @@ void shell_help(const struct shell *shell);
+ * @param[in] argv	Arguments.
+ * @param[in] ostr	String containing the legitimate option characters.
+ *
+ * @returns		If an option was successfully found, function returns
```
usually `@return` is used in zephyr.

nordic-krch commented on include/shell/shell_getopt.h:
```diff
@@ -0,0 +1,26 @@
+extern "C" {
+#endif
+
+/* Function initializing shell getopt module
```
missing `.`. Usually we don't start with `Funcion..` but write something more like `Initialize shell getopt module.`

nordic-krch reviewed (commented):
looks good, just a minor comments

carlescufi commented:
>Error: * scan/lib/getopt/getopt.c is not apache-2.0 licensed: bsd-new

@nashif @MaureenHelm @galak in-tree usage of a pre-approved license like BSD is acceptable, especially when it comes to a single file, is this right?

carlescufi commented:
> > Error: * scan/lib/getopt/getopt.c is not apache-2.0 licensed: bsd-new
> 
> @nashif @MaureenHelm @galak in-tree usage of a pre-approved license like BSD is acceptable, especially when it comes to a single file, is this right? `lib/fnmatch` set that precedent.

After a brief discussion, I think we should follow the procedure here:
https://docs.zephyrproject.org/latest/contribute/index.html?highlight=license#submission-and-review-process

@jakub-uC could you please add a README as described in the doc above?
I can then initiate the process to get TSC and board approval.

carlescufi commented on lib/getopt/getopt.c:
```diff
@@ -0,0 +1,128 @@
+ *	Parse argc/argv argument vector.
+ */
+int
+getopt(state, nargc, nargv, ostr)
```
@jakub-uC did you add this `state` parameter?

jakub-uC replied:
yes

Commit:
codeowners: add getopt library

Add owner for newly created getopt library.
Extend ownership of fnmatch library.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jakub-uC commented:
@carlescufi I've added the `README` file.

carlescufi commented on lib/getopt/getopt.c:
```diff
@@ -0,0 +1,128 @@
+/*	$NetBSD: getopt.c,v 1.26 2003/08/07 16:43:40 agc Exp $	*/
+
```
Could you please add an SPDX tag here?
```
SPDX-License-Identifier: BSD-3-Clause
```

jakub-uC replied:
done

carlescufi commented on lib/getopt/getopt.c:
```diff
@@ -0,0 +1,128 @@
+/*	$NetBSD: getopt.c,v 1.26 2003/08/07 16:43:40 agc Exp $	*/
+/* SPDX-License-Identifier: BSD-3-Clausei */
```
```suggestion
/* SPDX-License-Identifier: BSD-3-Clause */
```

jakub-uC replied:
@carlescufi fixed

martin-aberg commented on lib/getopt/getopt.c:
```diff
@@ -0,0 +1,128 @@
+	int nargc;
+	char *const nargv[];
+	const char *ostr;
+{
+	static char *place = EMSG;	/* option letter processing */
+	char *oli;			/* option letter list index */
```
Should `place` be in `state`?

jakub-uC replied:
Yes it should. I will fix that

jakub-uC replied:
fixed

nashif commented on CODEOWNERS:
```diff
@@ -453,8 +453,9 @@
-/lib/fnmatch/                             @carlescufi
+/lib/fnmatch/                             @carlescufi @jakub-uC
 /lib/gui/                                 @vanwinkeljan
+/lib/getopt/                              @jakub-uC
```
need to move this to some common place, having a lib per file (see fnmatch) is an overkill. Lets move all of those one lib level. This was discussed some time ago in the API meeting.

jakub-uC replied:
fixed

jakub-uC commented:
@nashif Should I create something like `lib/util` and put there fnmatch and getopt?

MaureenHelm commented:
> I can then initiate the process to get TSC and board approval.

The board didn't raise any objections, we can now move forward with this.

Commit:
lib: move fnmatch to lib/util

We expect to have more libraries with incopatible license. There must
be a common place for such software. It seems that lib/util is good
place for that.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
doc: shell update with getopt

Update of the shell documentation with getopt feature.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

carlescufi reviewed (approved):
`lib/util` sounds reasonable to me.

carlescufi commented:
@nashif can you take another look please? `lib/util` is vague, but I cannot come up with anything better myself.
